### PR TITLE
C++: Add support for more clang builtins

### PR DESCRIPTION
**Pull request by jketema**

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/exprs/Cast.qll [hunk 1]
```diff
@@ -791,6 +791,53 @@ class AlignofTypeOperator extends AlignofOperator {
...
+ */
+class DatasizeofExprOperator extends DatasizeofOperator {
+  DatasizeofExprOperator() { exists(this.getChild(0)) }
+
+  override string getAPrimaryQlClass() { result = "DatasizeofExprOperator" }
+
+  /** Gets the contained expression. */
+  Expr getExprOperand() { result = this.getChild(0) }
+
+  override string toString() { result = "sizeof(<expr>)" }
```
Comment (calumgrant) [inline cpp/ql/lib/semmle/code/cpp/exprs/Cast.qll:816]: I think this should be `__datasizeof`

File: cpp/ql/lib/semmle/code/cpp/exprs/Cast.qll [hunk 2]
```diff
@@ -791,6 +791,53 @@ class AlignofTypeOperator extends AlignofOperator {
...
+ */
+class DatasizeofTypeOperator extends DatasizeofOperator {
+  DatasizeofTypeOperator() { sizeof_bind(underlyingElement(this), _) }
+
+  override string getAPrimaryQlClass() { result = "DatasizeofTypeOperator" }
+
+  /** Gets the contained type. */
+  Type getTypeOperand() { sizeof_bind(underlyingElement(this), unresolveElement(result)) }
+
+  override string toString() { result = "sizeof(" + this.getTypeOperand().getName() + ")" }
```
Comment (calumgrant) [inline cpp/ql/lib/semmle/code/cpp/exprs/Cast.qll:834]: `__datasizeof`

File: cpp/ql/lib/semmle/code/cpp/exprs/Cast.qll [hunk 3]
```diff
@@ -791,6 +791,53 @@ class AlignofTypeOperator extends AlignofOperator {
   override string toString() { result = "alignof(" + this.getTypeOperand().getName() + ")" }
 }
 
+/**
+ * A C++ `__datasizeof` expression (used by some implementations
+ * of the `<type_traits>` header).
+ *
+ * The `__datasizeof` expression behaves identical to `sizeof` except
```
Comment (calumgrant) [inline cpp/ql/lib/semmle/code/cpp/exprs/Cast.qll:798]: Nit: identically

File: cpp/ql/test/library-tests/types/datasizeof/datasizeof.ql [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import cpp
+
+from DatasizeofOperator sto, string elemDesc, Element e
```
Comment (calumgrant) [inline cpp/ql/test/library-tests/types/datasizeof/datasizeof.ql:3]: Question: Would it be helpful to also output the constant value of the expression?

**Review by calumgrant — CHANGES_REQUESTED** [review]
Approach LGTM - would have expected `__datasizeof` rather than `sizeof` in the `toString()` but perhaps there is a good reason for this.

File: cpp/ql/test/library-tests/types/datasizeof/datasizeof.ql (hunk 4 above)
Comment (jketema) [reply cpp/ql/test/library-tests/types/datasizeof/datasizeof.ql:3]: Added. Also for the `sizeof` test from which this was copied.

**Review by calumgrant — APPROVED** [review]
Great!

**Pull request merged**